packages/ember-data/lib/serializers/json-serializer.js

```diff
@@ -274,11 +276,42 @@ export default Serializer.extend({
   normalizeErrors: function(typeClass, hash) {
     this.normalizeId(hash);
     this.normalizeAttributes(typeClass, hash);
-    this.normalizeRelationships(typeClass, hash);
+    this.normalizeRelationshipErrors(typeClass, hash);
     this.normalizeUsingDeclaredMapping(typeClass, hash);
   },
 
   /**
+    @method normalizeRelationshipErrors
+    @private
+  */
+  normalizeRelationshipErrors: function(typeClass, hash) {
+    if (!this.keyForRelationship) {
+      return;
+    }
+
+    typeClass.eachRelationship(function(key, relationship) {
+      var payloadKey = this.keyForRelationship(key, relationship.kind, 'deserialize');
+      if (key !== payloadKey && hash.hasOwnProperty(payloadKey)) {
+        hash[key] = hash[payloadKey];
+        delete hash[payloadKey];
+      }
+
+      if (!hash || !hash[key]) {
+        return;
+      }
```

Comment:
Actually, I'd like to avoid referencing anything embedded-record related in the JSON serializer. I just pushed what I think is a more robust implementation.